src/ripple/app/ledger/impl/BuildLedger.cpp

```diff
@@ -45,109 +45,97 @@ buildLedgerImpl(
-        int const tmf = buildLCL->txMap().flushDirty(
-            hotTRANSACTION_NODE, buildLCL->info().seq);
+        int const asf = built->stateMap().flushDirty(
+            hotACCOUNT_NODE, built->info().seq);
+        int const tmf = built->txMap().flushDirty(
+            hotTRANSACTION_NODE, built->info().seq);
         JLOG(j.debug()) << "Flushed " << asf << " accounts and " << tmf
                         << " transaction nodes";
     }
-    buildLCL->unshare();
+    built->unshare();
 
     // Accept ledger
-    buildLCL->setAccepted(
+    built->setAccepted(
         closeTime, closeResolution, closeTimeCorrect, app.config());
 
-    return buildLCL;
+    return built;
 }
 
 /** Apply a set of consensus transactions to a ledger.
 
   @param app Handle to application
-  @param txns Consensus transactions to apply
-  @param view Ledger to apply to
-  @param buildLCL Ledger to check if transaction already exists
+  @param txns the set of transactions to apply,
+  @param failed set of transactions that failed to apply
+  @param view ledger to apply to
   @param j Journal for logging
-  @return Any retriable transactions
+  @return number of transactions applied; transactions to retry left in txns
 */
 
-CanonicalTXSet
+std::size_t
 applyTransactions(
     Application& app,
-    SHAMap const& txns,
+    std::shared_ptr<Ledger const> const& built,
+    CanonicalTXSet& txns,
+    std::set<TxID>& failed,
     OpenView& view,
-    std::shared_ptr<Ledger> const& buildLCL,
     beast::Journal j)
 {
-    CanonicalTXSet retriableTxs(txns.getHash().as_uint256());
-
-    for (auto const& item : txns)
-    {
-        if (buildLCL->txExists(item.key()))
-            continue;
-
-        // The transaction wasn't filtered
-        // Add it to the set to be tried in canonical order
-        JLOG(j.debug()) << "Processing candidate transaction: " << item.key();
-        try
-        {
-            retriableTxs.insert(
-                std::make_shared<STTx const>(SerialIter{item.slice()}));
-        }
-        catch (std::exception const&)
-        {
-            JLOG(j.warn()) << "Txn " << item.key() << " throws";
-        }
-    }
-
     bool certainRetry = true;
+    std::size_t count = 0;
+
     // Attempt to apply all of the retriable transactions
     for (int pass = 0; pass < LEDGER_TOTAL_PASSES; ++pass)
     {
-        JLOG(j.debug()) << "Pass: " << pass << " Txns: " << retriableTxs.size()
-                        << (certainRetry ? " retriable" : " final");
+        JLOG(j.debug())
+            << (certainRetry ? "Pass: " : "Final pass: ") << pass
+            << " begins (" << txns.size() << " transactions)";
         int changes = 0;
 
-        auto it = retriableTxs.begin();
+        auto it = txns.begin();
 
-        while (it != retriableTxs.end())
+        while (it != txns.end())
         {
+            auto const txid = it->first.getTXID();
+
             try
             {
+                if (pass == 0 && built->txExists(txid))
```

Comment:
Is there a reason this filtering can't be done over in https://github.com/ripple/rippled/pull/2700/files#diff-256efa955bd129c0f8c25ae0f66d8afcR462 where `txns` is initially built?